File: 백준/Silver/1920. 수 찾기/수 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	
	static int N;
	static int[] arr;

	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();
		
		N = Integer.parseInt(br.readLine());
		arr = new int[N];
		st = new StringTokenizer(br.readLine());
		for (int i=0; i<N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		Arrays.sort(arr);
		
//		System.out.println(Arrays.toString(arr));
		
		int M = Integer.parseInt(br.readLine());
		st = new StringTokenizer(br.readLine());
		for (int i=0; i<M; i++) {
			sb.append(binarySearch(Integer.parseInt(st.nextToken()))+"\n");					// while 버전
//			sb.append(binarySearch(Integer.parseInt(st.nextToken()), 0, N-1)+"\n");		// 재귀 버전
		}
		
		System.out.println(sb);
		
		

	}
	
	static int binarySearch(int val) {
		
		int left = 0;
		int right = N-1;
		int mid = -1;
		
		while (left<=right) {
			
			mid = (left+right)/2;
			
			if (val < arr[mid]) {
				right = mid-1;
			} else  if (val > arr[mid]) {
				left = mid+1;
			} else {
				return 1;
			}
			
		}

		return 0;
	}
	
	
	
	
//	재귀버전
//	이분탐색은 재귀가 코드는 쉽지만 성능 측면에서 while문이 더 뛰어남 (스택을 쌓지 않으므로)
//	static int binarySearch(int val, int left, int right) { 
//		
//		if (right < left) {
//			return 0;
//		}
//		
//		int mid = (left+right)/2;
//		
////		System.out.println("val: " + val + " mid: " + arr[mid]);
//		
//		if (val < arr[mid]) {
//			return binarySearch(val, left, mid-1);
//		} else if (val > arr[mid]) {
//			return binarySearch(val, mid+1, right);
//		} else { // val == mid
//			return 1;
//		}
//			
//	}

}

File: 백준/Silver/1920. 수 찾기/수 찾기.py
def binary_search(t):
    s, e = 0, N - 1
    while s <= e:
        m = (s+e)//2
        if lst[m] == t:
            return 1        # 찾았다면 1 반환
        elif lst[m] < t:
            s = m+1         # m이 목표값보다 작다면 => s를 올려 조정 (m은 정답이 아니므로 미포함)
        else:   # lst[m] > t
            e = m-1         # m이 목표값보다 크다면 => e를 낮춰 조정 (m은 정답이 아니므로 미포함)
    return 0    # 못찾고 s>e 됐다면 종료


N = int(input())
lst = list(map(int, input().split()))
lst.sort() # 이진 탐색 사용을 위해서 정렬 필요

M = int(input())
query = list(map(int, input().split()))
for q in query:
    print(binary_search(q))


